lib/rpmsg/rpmsg_virtio.c

```diff
@@ -583,7 +581,15 @@ static void rpmsg_virtio_rx_callback(struct virtqueue *vq)
 		metal_mutex_acquire(&rdev->lock);
 
 		/* Check whether callback wants to hold buffer */
-		if (!(rp_hdr->reserved & RPMSG_BUF_HELD)) {
+		if (rp_hdr->reserved & RPMSG_BUF_HELD) {
+			/* Yes, save idx. */
+			rp_hdr->reserved |= idx;
+
+#ifdef VIRTIO_CACHED_BUFFERS
+			metal_cache_flush(rp_hdr, sizeof(struct rpmsg_hdr));
+#endif
```

Comment:
I'm a bit confused. Why do you need to flush the header here if the buffer is being held? I would expect that to be flushed when the buffer will be released and ready again for consumption.

Reply:
Because we modify the RX hdr data when hold the RX buffer.
Details please see this PR's description.

Reply:
The PR's description is really confusing to me. If we are holding the RX buffer that means that the buffer is being used by (for example) CPU1 and it will be released at a later time, so why it is important to flush the header at this stage?

Reply:
If we are holding the RX buffer that means that the buffer is being used by (for example) CPU1
Yes, this time we do:
3. set idx to hdr->reserved 
And this time, the hdr is cache-able for CPU1, the step 3 write action probably just write to cache (non write to real RAM), 
Then after step 5, the buffer returned to CPU0, but the CPU1 dirty cache (which caused by step 3 write action) auto flushed.
Then we meet error.

Reply:
The need seems valid to me,  but could not be sufficient. If the RPMsg client updates the rpmsg payload this could lead to  payload corruption on auto flush.
Seems to me that flushing the cache associated to the full buffer in `rpmsg_virtio_return_buffer` would be more reliable...